builder.md

```diff
@@ -1,57 +1,188 @@
+- _The standardized CLI spec_: This is a specification REQUIRED for
+  reference tooling where exposing a CLI is relevant. For non-reference tools,
+  all the clauses in this specification are to be considered optional.
 
-- `/templates/*.mustache` - A template file (there may be multiples of these)
-- `/templates/config.yaml` - A template configuration file
+With that said, builders are categorized into:
 
-## Workflow
-The first thing a builder needs to do is parse all the scheme files from the schemes repository (as defined in the [file guidelines](https://github.com/chriskempson/base16/blob/master/file.md)). All files matching the pattern `*.yaml` should be loaded from the root of the schemes repository.
+- _reference tooling_: These MUST follow the main spec and either expose a
+  command-line or a software library interface, for "plumbing" usage with
+  scripts or when building other, more complex, base16 tools. High quality
+  reference builders will usually be adopted by the base16-project
+  organization.
+- _extended tooling_: These are built for end users and are usually tailored
+  for a specific usecase. This means that they MUST follow _only_ the main
+  base16 (template variables) spec to attain full compliance. These include GUI
+  apps, ergonomical CLI tools, scheme managers, web apps, and more. Authors are
+  encouraged to use (or create) reference implementations as libraries, but may
+  also implement the builder feature themselves. Specs besides the template
+  variables are to be considered optional.
 
-When building a target template repository, a base16 builder should first clear out any old output. Next, for all templates defined in the template repo's config file (located at `/templates/config.yaml`), the builder should iterate through all the defined schemes and output matching files. The built filename should look like `[output-dir]/base16-[slug][extension]`, where the slug is taken from the scheme filename made lowercase with spaces replaced with dashes and both the extension and output-dir are taken from `/template/config.yaml`.
+## 2. Template variables spec
 
-In the case where schemes share the same slug, a builder will overwrite files previously generated from the template. Should this happen, a builder should show warning messages listing the overwritten files.
-
-## Template Variables
-A builder should provide the following variables to a template file:
+A builder tool MUST provide the following variables, and no others, to the
+template files it processes:
 
 - `scheme-name` - obtained from the scheme file
 - `scheme-author` - obtained from the scheme file
 - `scheme-slug` - obtained from the scheme filename, as described above
+- `scheme-slug-underscored` - same as `scheme-slug`, but with dashes replaced with underscores
+- `scheme-kind` - either "light" or "dark", calculated from the `base00` color <!-- TODO: This is a candidate for inclusion, let me know your thoughts -->
```

Comment:
it must have described the computation algorhythm for fallback values in older themes - otherwise it's again doesn't meet the promise of backwar-compatibility

Reply:
Huh? I'm not sure I follow.  I think we need to define what "backwards compatible" means... can you give examples of what you think we'd be "breaking" with these changes and how?

Reply:
means if:
1) scheme written for previous version of spec:
1.1) so it doesnt have `scheme-kind` field defined
2) template written for the current version of spec
2.1) so it have `scheme-kind` field used
3) builder must compute fallback value for `scheme-kind` field in such case and spec must define the logic (algorithm) for such computation

Reply:
i think the formula should be:
`"light" if (base00-r+base00-g+base00-b) > (base07-r+base07g+base07-b)else "dark"`

Reply:
> calculated from the `base00` color

Yes, I assume the spec would specify how this happens mathematically - no argument... there are existing luminescence formulas we should probably use.

Reply:
> there are existing luminescence formulas

in theory - yes, but on practice that would en-complicate the spec and so builder implementation notably, whilst r, g, b color computation is already defined in builder spec

Reply:
If we're talking about light or dark we should use a known luminescence formula - not reinvent the wheel.  I do not understand why you think using a common, known luminescence calculation would complicate the spec.